File: frontend/src/api/weather.ts
import { apiGet, apiPostEmpty, apiPut } from './http'
import {
  asArray,
  asBoolean,
  asNumber,
  asOptionalNumber,
  asOptionalString,
  asRecord,
  asString,
  type DisplayAlerts,
  type DisplayAlertPresentation,
  type DisplayWeather,
  type WeatherAlert,
  type WeatherAlertsState,
  type WeatherCurrentConditions,
  type WeatherLocation,
  type WeatherProviderState,
  type WeatherRefreshRun,
  type WeatherSettings,
  type WeatherStatus,
} from './types'

export async function getWeatherSettings(): Promise<WeatherSettings> {
  const payload = await apiGet<unknown>('/api/weather/settings')
  return normalizeWeatherSettings(payload)
}

export async function updateWeatherSettings(settings: WeatherSettings): Promise<WeatherSettings> {
  const payload = await apiPut<WeatherSettings, unknown>('/api/weather/settings', settings)
  return normalizeWeatherSettings(payload)
}

export async function getWeatherStatus(): Promise<WeatherStatus> {
  const payload = await apiGet<unknown>('/api/weather/status')
  const record = asRecord(payload)

  return {
    provider_status: normalizeWeatherProviderState(record?.provider_status),
    current_conditions: record?.current_conditions ? normalizeWeatherCurrentConditions(record.current_conditions) : null,
    dominant_alert: record?.dominant_alert ? normalizeWeatherAlert(record.dominant_alert) : null,
    active_alert_count: asNumber(record?.active_alert_count, 0),
    current_display_action: asString(record?.current_display_action, 'none'),
    recent_refresh_runs: asArray(record?.recent_refresh_runs, (item) => normalizeWeatherRefreshRun(item)),
  }
}

export async function getWeatherAlerts(): Promise<WeatherAlertsState> {
  const payload = await apiGet<unknown>('/api/weather/alerts')
  const record = asRecord(payload)

  return {
    provider_status: normalizeWeatherProviderState(record?.provider_status),
    alert_count: asNumber(record?.alert_count, 0),
    dominant_alert: record?.dominant_alert ? normalizeWeatherAlert(record.dominant_alert) : null,
    active_alerts: asArray(record?.active_alerts, (item) => normalizeWeatherAlert(item)),
  }
}

export async function refreshWeather(): Promise<WeatherStatus> {
  const payload = await apiPostEmpty<unknown>('/api/weather/refresh')
  const record = asRecord(payload)

  return {
    provider_status: normalizeWeatherProviderState(record?.provider_status),
    current_conditions: record?.current_conditions ? normalizeWeatherCurrentConditions(record.current_conditions) : null,
    dominant_alert: record?.dominant_alert ? normalizeWeatherAlert(record.dominant_alert) : null,
    active_alert_count: asNumber(record?.active_alert_count, 0),
    current_display_action: asString(record?.current_display_action, 'none'),
    recent_refresh_runs: asArray(record?.recent_refresh_runs, (item) => normalizeWeatherRefreshRun(item)),
  }
}

export async function getDisplayWeather(): Promise<DisplayWeather> {
  const payload = await apiGet<unknown>('/api/display/weather')
  const record = asRecord(payload)

  return {
    enabled: asBoolean(record?.enabled, false),
    position: normalizeWeatherPosition(record?.position),
    units: normalizeWeatherUnits(record?.units),
    show_precipitation: asBoolean(record?.show_precipitation, true),
    show_humidity: asBoolean(record?.show_humidity, true),
    show_wind: asBoolean(record?.show_wind, true),
    provider_status: normalizeWeatherProviderState(record?.provider_status),
    current_conditions: record?.current_conditions ? normalizeWeatherCurrentConditions(record.current_conditions) : null,
  }
}

export async function getDisplayAlerts(): Promise<DisplayAlerts> {
  const payload = await apiGet<unknown>('/api/display/alerts')
  const record = asRecord(payload)

  return {
    provider_status: normalizeWeatherProviderState(record?.provider_status),
    dominant_alert: record?.dominant_alert ? normalizeWeatherAlert(record.dominant_alert) : null,
    active_alerts: asArray(record?.active_alerts, (item) => normalizeWeatherAlert(item)),
    presentation: normalizeDisplayAlertPresentation(record?.presentation),
  }
}

function normalizeWeatherSettings(payload: unknown): WeatherSettings {
  const record = asRecord(payload)

  return {
    weather_enabled: asBoolean(record?.weather_enabled, false),
    weather_provider: asString(record?.weather_provider, 'nws'),
    weather_location: normalizeWeatherLocation(record?.weather_location),
    weather_units: normalizeWeatherUnits(record?.weather_units),
    weather_position: normalizeWeatherPosition(record?.weather_position),
    weather_refresh_minutes: asNumber(record?.weather_refresh_minutes, 15),
    weather_show_precipitation: asBoolean(record?.weather_show_precipitation, true),
    weather_show_humidity: asBoolean(record?.weather_show_humidity, true),
    weather_show_wind: asBoolean(record?.weather_show_wind, true),
    weather_alerts_enabled: asBoolean(record?.weather_alerts_enabled, true),
    weather_alert_fullscreen_enabled: asBoolean(record?.weather_alert_fullscreen_enabled, true),
    weather_alert_minimum_severity: normalizeWeatherSeverity(record?.weather_alert_minimum_severity),
    weather_alert_repeat_enabled: asBoolean(record?.weather_alert_repeat_enabled, true),
    weather_alert_repeat_interval_minutes: asNumber(record?.weather_alert_repeat_interval_minutes, 5),
    weather_alert_repeat_display_seconds: asNumber(record?.weather_alert_repeat_display_seconds, 20),
  }
}

function normalizeWeatherLocation(payload: unknown): WeatherLocation {
  const record = asRecord(payload)
  return {
    label: asString(record?.label, ''),
    latitude: asOptionalNumber(record?.latitude) ?? null,
    longitude: asOptionalNumber(record?.longitude) ?? null,
  }
}

function normalizeWeatherProviderState(payload: unknown): WeatherProviderState {
  const record = asRecord(payload)
  return {
    provider_name: asString(record?.provider_name, 'nws'),
    provider_display_name: asString(record?.provider_display_name, 'National Weather Service'),
    status: asString(record?.status, 'disabled'),
    available: asBoolean(record?.available, false),
    configured: asBoolean(record?.configured, false),
    location_label: asString(record?.location_label, ''),
    last_weather_refresh_at: asOptionalString(record?.last_weather_refresh_at) ?? null,
    last_alert_refresh_at: asOptionalString(record?.last_alert_refresh_at) ?? null,
    last_successful_weather_refresh_at: asOptionalString(record?.last_successful_weather_refresh_at) ?? null,
    last_successful_alert_refresh_at: asOptionalString(record?.last_successful_alert_refresh_at) ?? null,
    current_error: asOptionalString(record?.current_error) ?? null,
    updated_at: asString(record?.updated_at, ''),
  }
}

function normalizeWeatherCurrentConditions(payload: unknown): WeatherCurrentConditions {
  const record = asRecord(payload)
  return {
    provider_name: asString(record?.provider_name, 'nws'),
    provider_display_name: asString(record?.provider_display_name, 'National Weather Service'),
    location_label: asString(record?.location_label, ''),
    condition: asString(record?.condition, 'Unavailable'),
    icon_token: asString(record?.icon_token, 'cloudy'),
    temperature: asOptionalNumber(record?.temperature) ?? null,
    temperature_unit: asString(record?.temperature_unit, 'F'),
    humidity_percent: asOptionalNumber(record?.humidity_percent) ?? null,
    wind_speed: asOptionalNumber(record?.wind_speed) ?? null,
    wind_unit: asString(record?.wind_unit, 'mph'),
    wind_direction: asOptionalString(record?.wind_direction) ?? null,
    precipitation_probability_percent: asOptionalNumber(record?.precipitation_probability_percent) ?? null,
    observed_at: asOptionalString(record?.observed_at) ?? null,
    fetched_at: asString(record?.fetched_at, ''),
    attribution: asString(record?.attribution, 'National Weather Service'),
    is_stale: asBoolean(record?.is_stale, false),
  }
}

function normalizeWeatherAlert(payload: unknown): WeatherAlert {
  const record = asRecord(payload)
  return {
    id: asString(record?.id, ''),
    provider_name: asString(record?.provider_name, 'nws'),
    provider_display_name: asString(record?.provider_display_name, 'National Weather Service'),
    event: asString(record?.event, 'Weather Alert'),
    severity: asString(record?.severity, 'unknown'),
    certainty: asString(record?.certainty, 'Unknown'),
    urgency: asString(record?.urgency, 'Unknown'),
    headline: asString(record?.headline, 'Weather Alert'),
    description: asString(record?.description, ''),
    instruction: asString(record?.instruction, ''),
    area: asString(record?.area, ''),
    status: asString(record?.status, 'Actual'),
    issued_at: asOptionalString(record?.issued_at) ?? null,
    effective_at: asOptionalString(record?.effective_at) ?? null,
    expires_at: asOptionalString(record?.expires_at) ?? null,
    ends_at: asOptionalString(record?.ends_at) ?? null,
    attribution: asString(record?.attribution, 'National Weather Service'),
    escalation_mode: asString(record?.escalation_mode, 'badge'),
    effective_escalation_mode: asString(record?.effective_escalation_mode, asString(record?.escalation_mode, 'badge')),
    display_priority: asNumber(record?.display_priority, 0),
    effective_display_priority: asNumber(record?.effective_display_priority, 0),
    event_priority: asNumber(record?.event_priority, 0),
    is_active: asBoolean(record?.is_active, true),
    is_dominant: asBoolean(record?.is_dominant, false),
  }
}

function normalizeWeatherRefreshRun(payload: unknown): WeatherRefreshRun {
  const record = asRecord(payload)
  return {
    id: asString(record?.id, ''),
    provider_name: asString(record?.provider_name, 'nws'),
    refresh_kind: asString(record?.refresh_kind, 'weather'),
    trigger: asString(record?.trigger, 'manual'),
    status: asString(record?.status, 'unknown'),
    message: asString(record?.message, ''),
    error_message: asOptionalString(record?.error_message) ?? null,
    started_at: asString(record?.started_at, ''),
    completed_at: asOptionalString(record?.completed_at) ?? null,
  }
}

function normalizeDisplayAlertPresentation(payload: unknown): DisplayAlertPresentation {
  const record = asRecord(payload)
  return {
    mode: asString(record?.mode, 'none'),
    fallback_mode: asOptionalString(record?.fallback_mode) ?? null,
    repeat_interval_minutes: asNumber(record?.repeat_interval_minutes, 5),
    repeat_display_seconds: asNumber(record?.repeat_display_seconds, 20),
    alert_count: asNumber(record?.alert_count, 0),
  }
}

function normalizeWeatherPosition(value: unknown): DisplayWeather['position'] {
  switch (value) {
    case 'top-left':
    case 'top-right':
    case 'bottom-left':
    case 'bottom-right':
      return value
    default:
      return 'top-right'
  }
}

function normalizeWeatherUnits(value: unknown): DisplayWeather['units'] {
  return value === 'c' ? 'c' : 'f'
}

function normalizeWeatherSeverity(value: unknown): WeatherSettings['weather_alert_minimum_severity'] {
  switch (value) {
    case 'unknown':
    case 'minor':
    case 'moderate':
    case 'severe':
    case 'extreme':
      return value
    default:
      return 'minor'
  }
}

File: backend/app/schemas/weather.py
from __future__ import annotations

from dataclasses import asdict
from typing import Literal

from pydantic import BaseModel, Field, model_validator

from app.models.weather import WeatherLocation, WeatherSettings


class WeatherLocationPayload(BaseModel):
    label: str = Field(default="", max_length=120)
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)

    @model_validator(mode="after")
    def require_both_coordinates_or_neither(self) -> "WeatherLocationPayload":
        if (self.latitude is None) != (self.longitude is None):
            raise ValueError("weather_location.latitude and weather_location.longitude must either both be set or both be empty")
        return self

    @classmethod
    def from_domain(cls, location: WeatherLocation) -> "WeatherLocationPayload":
        return cls(label=location.label, latitude=location.latitude, longitude=location.longitude)


class WeatherSettingsResponse(BaseModel):
    weather_enabled: bool
    weather_provider: str
    weather_location: WeatherLocationPayload
    weather_units: Literal["f", "c"]
    weather_position: Literal["top-left", "top-right", "bottom-left", "bottom-right"]
    weather_refresh_minutes: int
    weather_show_precipitation: bool
    weather_show_humidity: bool
    weather_show_wind: bool
    weather_alerts_enabled: bool
    weather_alert_fullscreen_enabled: bool
    weather_alert_minimum_severity: Literal["unknown", "minor", "moderate", "severe", "extreme"]
    weather_alert_repeat_enabled: bool
    weather_alert_repeat_interval_minutes: int
    weather_alert_repeat_display_seconds: int

    @classmethod
    def from_domain(cls, settings: WeatherSettings) -> "WeatherSettingsResponse":
        payload = asdict(settings)
        payload["weather_location"] = WeatherLocationPayload.from_domain(settings.weather_location)
        return cls(**payload)


class WeatherSettingsUpdateRequest(BaseModel):
    weather_enabled: bool
    weather_provider: str = Field(min_length=1, max_length=60)
    weather_location: WeatherLocationPayload
    weather_units: Literal["f", "c"]
    weather_position: Literal["top-left", "top-right", "bottom-left", "bottom-right"]
    weather_refresh_minutes: int = Field(ge=1, le=180)
    weather_show_precipitation: bool
    weather_show_humidity: bool
    weather_show_wind: bool
    weather_alerts_enabled: bool
    weather_alert_fullscreen_enabled: bool
    weather_alert_minimum_severity: Literal["unknown", "minor", "moderate", "severe", "extreme"]
    weather_alert_repeat_enabled: bool
    weather_alert_repeat_interval_minutes: int = Field(ge=1, le=120)
    weather_alert_repeat_display_seconds: int = Field(ge=5, le=300)

    @model_validator(mode="after")
    def require_location_when_enabled(self) -> "WeatherSettingsUpdateRequest":
        if self.weather_enabled and (self.weather_location.latitude is None or self.weather_location.longitude is None):
            raise ValueError("weather_location.latitude and weather_location.longitude are required when weather is enabled")
        return self

    def to_domain(self) -> WeatherSettings:
        return WeatherSettings(
            weather_enabled=self.weather_enabled,
            weather_provider=self.weather_provider,
            weather_location=WeatherLocation(
                label=self.weather_location.label,
                latitude=self.weather_location.latitude,
                longitude=self.weather_location.longitude,
            ),
            weather_units=self.weather_units,
            weather_position=self.weather_position,
            weather_refresh_minutes=self.weather_refresh_minutes,
            weather_show_precipitation=self.weather_show_precipitation,
            weather_show_humidity=self.weather_show_humidity,
            weather_show_wind=self.weather_show_wind,
            weather_alerts_enabled=self.weather_alerts_enabled,
            weather_alert_fullscreen_enabled=self.weather_alert_fullscreen_enabled,
            weather_alert_minimum_severity=self.weather_alert_minimum_severity,
            weather_alert_repeat_enabled=self.weather_alert_repeat_enabled,
            weather_alert_repeat_interval_minutes=self.weather_alert_repeat_interval_minutes,
            weather_alert_repeat_display_seconds=self.weather_alert_repeat_display_seconds,
        )


class WeatherProviderStateResponse(BaseModel):
    provider_name: str
    provider_display_name: str
    status: str
    available: bool
    configured: bool
    location_label: str
    last_weather_refresh_at: str | None
    last_alert_refresh_at: str | None
    last_successful_weather_refresh_at: str | None
    last_successful_alert_refresh_at: str | None
    current_error: str | None
    updated_at: str


class WeatherCurrentConditionsResponse(BaseModel):
    provider_name: str
    provider_display_name: str
    location_label: str
    condition: str
    icon_token: str
    temperature: int | None
    temperature_unit: str
    humidity_percent: int | None
    wind_speed: int | None
    wind_unit: str
    wind_direction: str | None
    precipitation_probability_percent: int | None
    observed_at: str | None
    fetched_at: str
    attribution: str
    is_stale: bool


class WeatherAlertResponse(BaseModel):
    id: str
    provider_name: str
    provider_display_name: str
    event: str
    severity: str
    certainty: str
    urgency: str
    headline: str
    description: str
    instruction: str
    area: str
    status: str
    issued_at: str | None
    effective_at: str | None
    expires_at: str | None
    ends_at: str | None
    attribution: str
    escalation_mode: str
    effective_escalation_mode: str
    display_priority: int
    effective_display_priority: int
    event_priority: int
    is_active: bool
    is_dominant: bool


class WeatherRefreshRunResponse(BaseModel):
    id: str
    provider_name: str
    refresh_kind: str
    trigger: str
    status: str
    message: str
    error_message: str | None
    started_at: str
    completed_at: str | None


class WeatherStatusResponse(BaseModel):
    provider_status: WeatherProviderStateResponse
    current_conditions: WeatherCurrentConditionsResponse | None
    dominant_alert: WeatherAlertResponse | None
    active_alert_count: int
    current_display_action: str
    recent_refresh_runs: list[WeatherRefreshRunResponse]


class WeatherAlertsResponse(BaseModel):
    provider_status: WeatherProviderStateResponse
    alert_count: int
    dominant_alert: WeatherAlertResponse | None
    active_alerts: list[WeatherAlertResponse]


class DisplayWeatherResponse(BaseModel):
    enabled: bool
    position: Literal["top-left", "top-right", "bottom-left", "bottom-right"]
    units: Literal["f", "c"]
    show_precipitation: bool
    show_humidity: bool
    show_wind: bool
    provider_status: WeatherProviderStateResponse
    current_conditions: WeatherCurrentConditionsResponse | None


class DisplayAlertPresentationResponse(BaseModel):
    mode: str
    fallback_mode: str | None
    repeat_interval_minutes: int
    repeat_display_seconds: int
    alert_count: int


class DisplayAlertsResponse(BaseModel):
    provider_status: WeatherProviderStateResponse
    dominant_alert: WeatherAlertResponse | None
    active_alerts: list[WeatherAlertResponse]
    presentation: DisplayAlertPresentationResponse
